tls/s2n_client_hello.c

```diff
@@ -348,17 +351,20 @@ static S2N_RESULT s2n_client_hello_process_cb_response(struct s2n_connection *co
 
 int s2n_client_hello_recv(struct s2n_connection *conn)
 {
-    POSIX_ENSURE(conn->client_hello.callback_async_blocked == 0, S2N_ERR_ASYNC_BLOCKED);
+    if (conn->client_hello.callback_enable_poll == 0) {
+        POSIX_ENSURE(conn->client_hello.callback_async_blocked == 0, S2N_ERR_ASYNC_BLOCKED);
+    }
 
     if (conn->config->client_hello_cb_mode == S2N_CLIENT_HELLO_CB_BLOCKING ||
```

Comment:
Probably the best long-term approach to this logic should be to add a new "polling" callback mode to the existing s2n_client_hello_cb_mode options: https://github.com/aws/s2n-tls/blob/6df5dbfd96b20e1179b1291ab916f4a710b3d62e/api/s2n.h#L322 Particularly because blocking mode + polling doesn't make any sense. However, that would change the public API and I think you're trying to avoid that, so I guess this is fine for now.